application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -203,11 +219,100 @@ public void updateData(String fromVersion) throws Exception {
                     log.info("Skipping edge events migration");
                 }
                 break;
+            case "3.5.0":
+                log.info("Updating data from version 3.5.0 to 3.5.1 ...");
+                upgradeRuleNodes();
+                break;
             default:
                 throw new RuntimeException("Unable to update data, unsupported fromVersion: " + fromVersion);
         }
     }
 
+    private void upgradeRuleNodes() {
+        var ruleChainIdToTenantIdMap = new HashMap<RuleChainId, TenantId>();
+        try {
+            log.info("Starting rule nodes upgrade ...");
+            var ruleNodeDefinitions = getBeanDefinitions(
+                    org.thingsboard.rule.engine.api.RuleNode.class
+            );
+            for (BeanDefinition def : ruleNodeDefinitions) {
+                String clazzName = def.getBeanClassName();
+                Class<?> clazz = Class.forName(clazzName);
+                TbNode tbNode = (TbNode) clazz.getDeclaredConstructor().newInstance();
+                if (tbNode instanceof VersionedNode) {
+                    var versionedNode = (VersionedNode) tbNode;
+                    var ruleNodeName = versionedNode.getClass().getName();
+                    int currentVersion = versionedNode.getCurrentVersion();
+                    var ruleNodesToUpdate = new PageDataIterable<>(
+                            pageLink ->
+                                    ruleChainService.findAllRuleNodesByTypeAndVersionLessThan(ruleNodeName, currentVersion, pageLink),
+                            1024
+                    );
+                    for (RuleNode ruleNode : ruleNodesToUpdate) {
+                        RuleNodeId ruleNodeId = ruleNode.getId();
+                        var oldConfiguration = ruleNode.getConfiguration();
+                        int fromVersion = ruleNode.getConfigurationVersion();
+                        log.info("Going to upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {}",
+                                ruleNodeId,
+                                clazzName,
+                                fromVersion,
+                                currentVersion);
+                        try {
+                            TbPair<Boolean, JsonNode> upgradeRuleNodeConfigurationResult = versionedNode.upgrade(ruleNodeId, fromVersion, oldConfiguration);
+                            if (upgradeRuleNodeConfigurationResult.getFirst()) {
+                                ruleNode.setConfiguration(upgradeRuleNodeConfigurationResult.getSecond());
+                                var ruleChainId = ruleNode.getRuleChainId();
+                                var tenantId = getTenantId(ruleChainIdToTenantIdMap, ruleNodeId, ruleChainId);
```

Comment:
So that you know, you don't need the tenantId here. You may use the system tenant id. The tenantId in the dao is used only for the rate limits here.